src/client/api/job.c

```diff
@@ -0,0 +1,112 @@
+/*
+ * (C) Copyright 2020 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. 8F-30005.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+
+#include <stdlib.h>
+#include <sys/utsname.h>
+#include <daos/common.h>
+#include <daos/job.h>
+
+
+/*
+ * The answer of what the max length of envvar name is very tricky. Arguments
+ * and environment variable share the same memory space so to make things easy
+ * we enforce an arbitrary length of 80 which some other shells enforce
+ */
+#define MAX_ENV_NAME 80
+
+#define MAX_JOBID_LEN 1024
+
+char *dc_jobid_env;
+char *dc_jobid;
+
+static int craft_default_jobid(char **jobid)
+{
+	struct utsname name = {0};
+	pid_t pid;
+	int ret;
+
+	ret = uname(&name);
+	if (ret)
+		return -DER_INVAL;
```

Comment:
- IMO we should log this failure with whatever errno is set, since it'd be weird for uname to fail.
- For the return value, there is a DAOS function to translate an errno to a DER value.
- Reading all the notes on the uname manpage, I'm wondering... Any reason not to use gethostname instead?

Reply:
Really the only way for uname to fail is if an invalid buffer is passed into the call. Since we define the uname struct right there we might not even need to check the return code here at all since we know its a valid buffer. The only errno it can return is EFAULT so maybe we just convert it to the corresponding DER value like you suggest.

I didn't use gethostname becuase the interface isn't as easy to use and it uses uname under the hood anyway. With uname we have fixed array sizes we can use for a max length memcpy. With gethostname I have to guess at the hostname length and allocate that buffer and hope its large enough and if it isn't either accept the truncated hostname or realloc the buffer and try again.

Reply:
Ah, okay. I had thought gethostname looked a little more straightforward, but it sounds like not. But yeah I still think we should log and convert the actual errno if it fails, instead of using DER_INVAL, which I think could be confusing.